tests/safe_command/test_injection.py

```diff
@@ -0,0 +1,674 @@
+import pytest
+import subprocess
+from pathlib import Path
+from os import mkfifo, symlink, remove, getenv
+from shutil import rmtree, which
+
+from security import safe_command
+from security.safe_command.api import _parse_command, _resolve_paths_in_parsed_command, _shell_expand
+from security.exceptions import SecurityException
+
+with (Path(__file__).parent / "fuzzdb" / "command-injection-template.txt").open() as f:
+    FUZZDB_OS_COMMAND_INJECTION_PAYLOADS = [line.replace('\\n','\n').replace("\\'", "'")[:-1] for line in f] # Remove newline
+with (Path(__file__).parent / "fuzzdb" / "traversals-8-deep-exotic-encoding.txt").open() as f:
+    FUZZDB_PATH_TRAVERSAL_PAYLOADS = [line.replace('\\n','\n').replace("\\'", "'")[:-1] for line in f] # Remove newline
+
+
+@pytest.fixture
+def setup_teardown(tmpdir):
+    # Working directory is the tmpdir
+    wd = Path(tmpdir)
+    wd.mkdir(exist_ok=True)
+
+    # Create some files and directories to use in the tests
+    testtxt = wd / "test.txt"
+    testtxt.write_text("USERDATA1\nUSERDATA2\nUSERDATA3\n")
+    test2txt = wd / "test2.txt"
+    test2txt.write_text("USERDATA4\nUSERDATA5\nUSERDATA6\n")
+    rglob_testdir = wd / "rglob_testdir"
+    rglob_testdir.mkdir()
+    rglob_testfile = rglob_testdir / "rglob_testfile.txt"
+    rglob_testfile.touch()
+    space_in_name = wd / "space in name.txt"
+    space_in_name.touch()
+
+    testtxt.touch()
+    test2txt.touch()
+    cwd_testfile = Path("./cwdtest.txt").resolve()
+    cwd_testfile.touch()
+    fifo_testfile = (wd / "fifo_testfile").resolve()
+    mkfifo(fifo_testfile)
+    symlink_testfile = (wd / "symlink_testfile").resolve()
+    symlink(cwd_testfile, symlink_testfile) # Target of symlink_testfile is cwd_testfile.txt
+    passwd = Path("/etc/passwd").resolve()
+    sudoers = Path("/etc/sudoers").resolve()
+    # Get Path objects for the test commands
+    cat, echo, grep, nc, curl, sh = map(lambda cmd: Path(which(cmd) or f"/usr/bin/{cmd}" ), ["cat", "echo", "grep", "nc", "curl", "sh"]) 
+    testpaths = {
+        "wd": wd,
+        "test.txt": testtxt,
+        "test2.txt": test2txt,
+        "rglob_testdir": rglob_testdir,
+        "rglob_testfile": rglob_testfile,
+        "space_in_name": space_in_name,
+        "cwd_testfile": cwd_testfile,
+        "fifo_testfile": fifo_testfile,
+        "symlink_testfile": symlink_testfile,
+        "passwd": passwd,
+        "sudoers": sudoers,
+        "cat": cat,
+        "echo": echo,
+        "grep": grep,
+        "nc": nc,
+        "curl": curl,
+        "sh": sh
+    }
+    yield testpaths
+    
+    # Clean up the test files and directories
```

Comment:
Is this strictly necessary if it's already managed by the `tmpdir` fixture?

Reply:
You're right `rmtree(tmpdir, ignore_errors=True)` is redundant and can be removed since it is handled by the `tmpdir` fixture.

However `remove(cwd_testfile)` cannot be removed since it is needed to cleanup the `cwd_testfile` which is not in `tmpdir`. The purpose of this file is to test that everything works correctly when referencing the relative path of a file in the current directory, and it also double as the target of the symlink test.